build.gradle.kts

```diff
@@ -0,0 +1,663 @@
+        // In fact, for X < Y it's perfectly possible to compile Java X code that uses Java Y APIs...
+        // This will work fine, until we actually try to run those compiled classes under Java X-compatible JVM,
+        // when we'll end up with NoSuchMethodError for APIs added between Java X and Java Y
+        // (i.e. for X=8 and Y=11: InputStream#readAllBytes, Stream#takeWhile and String#isBlank).
+        // `options.release = X` makes sure that regardless of Java version used to run the compiler,
+        // only Java X-compatible APIs are available to the compiled code.
+        // options.release = Integer.parseInt(javaMajorVersion.majorVersion)
+        options.compilerArgs.addAll(listOf("--release", javaMajorVersion.majorVersion))
+    }
+
+    tasks.withType<Javadoc> {
+        // See JDK-8200363 (https://bugs.openjdk.java.net/browse/JDK-8200363) for information about the `-Xwerror` option:
+        // this is needed to make sure that javadoc always fails on warnings
+        // (esp. important on CI since javadoc there for some reason seems to never raise any errors otherwise).
+
+        // The '-quiet' as second argument is actually a hack around https://github.com/gradle/gradle/issues/2354:
+        // since the one-parameter `addStringOption` doesn't seem to work, we need to add an extra `-quiet`, which is added anyway by Gradle.
+        // options.addStringOption("Xwerror", "-quiet")
+        // options.addStringOption("Xdoclint:all", "-quiet")
+        options.quiet()
+    }
+
+    tasks.withType<Test> {
+        testLogging {
+            events = setOf(TestLogEvent.FAILED)
+            exceptionFormat = TestExceptionFormat.FULL
+            showCauses = true
+            showExceptions = true
+            showStackTraces = true
+        }
+    }
+
+
+    apply<SpotlessPlugin>()
+    configure<SpotlessExtension> {
+        java {
+            importOrder("java", "javax", "", "com.virtuslab")
+            // See https://github.com/diffplug/spotless/blob/master/ECLIPSE_SCREENSHOTS.md on importing and exporting settings from Eclipse
+            eclipse().configFile("$rootDir/config/spotless/formatting-rules.xml")
+            removeUnusedImports()
+            targetExclude("**/build/generated/**/*.*")
+        }
+
+        kotlin {
+            ktfmt()
+        }
+        scala {
+            scalafmt().configFile("$rootDir/scalafmt.conf")
+        }
+        groovy {
+            target("/buildSrc/**/*.groovy")
+            greclipse().configFile("$rootDir/config/spotless/formatting-rules.xml")
+        }
+    }
+
+    if (!isCI) {
+        tasks.withType<JavaCompile> {
+            dependsOn("spotlessJavaApply")
+        }
+        tasks.withType<KotlinCompile> {
+            dependsOn("spotlessKotlinApply")
+        }
+        tasks.withType<ScalaCompile> {
+            dependsOn("spotlessScalaApply")
+        }
+        tasks.withType<GroovyCompile> {
+            dependsOn("spotlessGroovyApply")
```

Comment:
Oh! so Groovy code in buildSrc to be migrated separately? or on this PR as well

Reply:
Separate - #1004